### PR TITLE
Not found handler. Registerer.

### PR DESCRIPTION
**Pull request by ryanvoong**

### CODE REVIEW

File: src/request_handler.cc [hunk 1]
```diff
@@ -1,111 +1,17 @@
...
 
-// todo: put each request handler into a separate file
+std::map<std::string, RequestHandler* (*)(void)>* request_handler_builders = nullptr;
 
-const std::string CONTENT_TYPE = "Content-Type";
-const std::string TEXT_PLAIN = "text/plain";
-const std::string STATIC_CONFIG_ROOT_KEY = "root";
+RequestHandler* RequestHandler::CreateByName(const char* type) {
+  const auto type_and_builder = request_handler_builders->find(type);
+  if (type_and_builder == request_handler_builders->end()) {
```
Comment (crystalhsieh) [inline src/request_handler.cc:7]: Wasn't it said previously that count()>0 would be preferred over checking for end() ? If not, that's okay too.

File: src/static_handler.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix,
+                                           const NginxConfig& config) {
+  uri_prefix_ = uri_prefix;
+  RequestHandler::ParseConfig(config);
+  root_ = config_map_[CONFIG_ROOT_KEY];
+  return OK;
+}
+
+// todo: add request validation
```
Comment (crystalhsieh) [inline src/static_handler.cc]: please change to capital TODO so it's easier to find.

File: src/static_handler.cc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+RequestHandler::Status StaticHandler::HandleRequest(const Request& request,
+                                                    Response* response) {
+  if (response == nullptr) {
+    return ERROR;
+  }
+
+  // todo: change how we get relative_path
```
Comment (crystalhsieh) [inline src/static_handler.cc]: ..? Is this currently looking for a uri prefix in static handler that matches the map, and if it doesnt, you reduce your size by 1?

File: src/static_handler.cc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    relative_path_string = root_ + uri.substr(uri_prefix_.size() - 1);
+  } else {
+    return ERROR;
+  }
+
+  boost::filesystem::path relative_path(relative_path_string);
+
+  if (boost::filesystem::exists(relative_path)
+      && boost::filesystem::is_regular_file(relative_path)) {
+    boost::filesystem::path absolute_path = boost::filesystem::canonical(relative_path);
```
Comment (crystalhsieh) [inline src/static_handler.cc]: please reformat so it's on a second line

File: src/echo_handler.cc [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#include "echo_handler.h"
+
+RequestHandler::Status EchoHandler::Init(const std::string& /*uri_prefix*/,
+                                         const NginxConfig& /*config*/) {
+  return OK;
+}
+
+// todo: check if request.uri() matches uri_prefix_ ?
```
Comment (crystalhsieh) [inline src/echo_handler.cc:8]: does the echo handler have any paths/uri's to check .. ?

File: src/request_handler.cc (hunk 1 above)
Comment (ryanvoong) [reply src/request_handler.cc:7]: This was copy/pasted from the profs. It doesn't matter if you use `count` or `end`, but I just think that `count` is cleaner

File: src/echo_handler.cc (hunk 5 above)
Comment (ryanvoong) [reply src/echo_handler.cc:8]: it depends if the URI checking is done by `Connection` or `EchoHandler`

File: src/static_handler.cc (hunk 2 above)
Comment (ryanvoong) [reply src/static_handler.cc]: how is it easier to find? can't you just ctrl-F "todo"

File: src/static_handler.cc (hunk 3 above)
Comment (ryanvoong) [reply src/static_handler.cc]: it's getting the relative file path by taking the request URI, removing the StaticHandler's URI prefix, and prepending the root path. I added a comment describing what's going on

File: src/static_handler.cc (hunk 4 above)
Comment (ryanvoong) [reply src/static_handler.cc]: done

**Pull request merged**